### PR TITLE
promote_type -> promote_op

### PR DESCRIPTION
**Pull request by lkdvos**

### CODE REVIEW

File: src/fallbacks.jl [hunk 1]
```diff
@@ -204,7 +207,7 @@ function add!!(y, x, α::Number, β::Number)
         Tx = scalartype(x)
         Ty = scalartype(y)
         if applicable(LinearAlgebra.axpby!, α′, x, β, y) &&
-           promote_type(typeof(α), Tx, typeof(β), Ty) <: Ty
+           Base.promote_op(add, Ty, Tx, scalartype(α), scalartype(β)) <: Ty
```
Comment (Jutho) [inline src/fallbacks.jl]: Above you kept the `typeof` for the scalar factors. I have no preference between `scalartype` and `typeof` for those, should be equivalent. `typeof` might be slightly shorter.

**Comment by Jutho** [comment]
Is this ready to merge (aside from the fact that it needs another format run)?

File: src/abstractarray.jl [hunk 2]
```diff
@@ -70,24 +69,25 @@ end
 
 # Special case: simple numerical arrays with BLAS-compatible floating point type
 function add!(y::BLASVector{T}, x::BLASVector{T},
-              α::Number=One(), β::Number=One()) where {T<:BlasFloat}
+              α::Number, β::Number) where {T<:BlasFloat}
+    α′ = α === One() ? true : convert(T, α)
```
Comment (Jutho) [inline src/abstractarray.jl]: I don't think that line is necessary? The true BLAS methods, to which `LinearAlgebra.axpy!` should lower, require all scalar types to be the same, so `true` is anyway also converted to type `T`.

File: src/interface.jl [hunk 3]
```diff
@@ -99,17 +99,24 @@ Add `y` and `x`, or more generally construct the linear combination `y * β + x
...
-    add!(y, x, [α::Number = 1, β::Number = 1])
+    add!(y, x, [α::Number = One(), β::Number = One()])
 
 Add `y` and `x`, or more generally construct the linear combination `y * β + x * α`, storing
 the result in `y`. This will error in case of incompatible scalar types or incommensurate sizes.
 
+!!! note
+    New types should only implement the four-argument version. When desired, the two- and
+    three-argument version can be distinguished by specializing on `α` and `β` being
+    `One()`.
```
Comment (Jutho) [inline src/interface.jl]: alpha and beta "being of type `One`" ?

File: src/abstractarray.jl (hunk 2 above)
Comment (lkdvos) [reply src/abstractarray.jl:73]: Ah, I didn't know this is how BLAS behaves, I'll fix this

Comment (lkdvos) [reply src/abstractarray.jl:73]: I didn't know of this behaviour for BLAS. I guess for the fallbacks we still want to keep the conversion, as this somehow assumes that VectorInterface has not been implemented, and these methods thus might not know about `One`?

**Comment by lkdvos** [comment]
Let's wait just a bit longer before merging, I will locally test the downstream packages and see if there is nothing I missed, as I did not really add a lot of tests for the new `Zero` and `One`. I guess another formatting run and a version update are necessary anyways

File: src/abstractarray.jl (hunk 2 above)
Comment (lkdvos) [reply src/abstractarray.jl]: I updated this here, but left it in the fallbacks, as presumably these methods do not know about the existence of `One` and `Zero`

Comment (Jutho) [reply src/abstractarray.jl]: Indeed, there it is different, and we just preserve alpha unless it is of type One.

**Comment by lkdvos** [comment]
This should now be ready for merging/registering a new version
